homeassistant/components/tfl/sensor.py

```diff
@@ -0,0 +1,184 @@
+            )
+            devices.append(
+                StopPointSensor(
+                    stop_point_api,
+                    __getCommonName(stop_point_id, stop_point_info),
+                    stop_point_id,
+                    unique_id,
+                )
+            )
+        except HTTPError as exception:
+            error_code = exception.code
+            _LOGGER.exception(
+                "Error retrieving stop point info from TfL for stop_point=%s with HTTP error_code=%s, entity will not be created for this stop point",
+                stop_point_ids,
+                error_code,
+            )
+        except URLError as exception:
+            _LOGGER.exception(
+                "Error retrieving stop point info from TfL for stop_point=%s with URLError reason=%s, entity will not be created for this stop point",
+                stop_point_ids,
+                exception.reason,
+            )
+
+    async_add_entities(devices, True)
+
+
+def __getCommonName(stop_point_id: str, stop_point_info: dict[str, typing.Any]) -> str:
+    """Get the common name for the Stop Point.
+
+    In the happy path, the Stop Point Id will match the naptanId, and this is simple.
+    For Stop Points where there are many stops that can be grouped together, the TfL API
+    will group them and we need to find our target Stop Point Id and get the common name.
+    """
+    if stop_point_info["naptanId"] == stop_point_id:
+        return typing.cast(str, stop_point_info["commonName"])
+
+    # The stop point is more complicated and hierarchical, and we need to find the one that was asked for
+    for child_stop_point in stop_point_info["children"]:
+        for subchild in child_stop_point["children"]:
+            if subchild["naptanId"] == stop_point_id:
+                return typing.cast(str, subchild["commonName"])
+    return "Unknown"
+
+
+class StopPointSensor(SensorEntity):
+    """Representation of a TfL StopPoint as a Sensor.
+
+    The Sensor does not use a DataUpdateCoordinator because
+    TfL doesn't have an API to query multiple stops in a single call. There's a 1:1 mapping between a sensor and an API call.
+    """
+
+    _attr_attribution = "Powered by TfL Open Data"
+    _attr_icon = "mdi:bus"
+    _attr_has_entity_name = True
+
+    def __init__(
+        self, stop_point_api: stopPoint, name: str, stop_point_id: str, unique_id: str
+    ) -> None:
+        """Initialize the TfL StopPoint sensor."""
+        self._name = name
+        self._attr_name = name
+        self._attr_unique_id = f"{unique_id}_{stop_point_id}"
+        self._attr_state_class = SensorStateClass.MEASUREMENT
+        self._attr_device_info = DeviceInfo(
+            name="TfL",
+            identifiers={(DOMAIN, unique_id)},
+            entry_type=DeviceEntryType.SERVICE,
+        )
+
+        self._stop_point_api = stop_point_api
+        self._stop_point_id = stop_point_id
+
+    async def async_update(self) -> None:
```

Comment:
Ah, I see how that works. Yes, that is quite a nice pattern. Thanks for referencing the example.
I can make that refactor, a little bigger than I was hoping for but that's alright

Reply:
Hmm, actually it's more complex. The `swiss_public_transport` integration is doing a point A to point B connection, so there is no additional information associated with each connection.

That's not the case for a TfL stop point. There may be say 3 different buses going through a stop point which all go in the same direction up to a point but have different final destinations. Similarly, for a tube / metro stop point there may be 2 different lines running through a stop point (a tube station platform) going in the same direction up to a point; or the same tube line but with different destinations (either not going to the end of the line or some tube lines have multiple destinations).

@joostlek how would you recommend encoding that information? I could create multiple sensors for each arrival, one sensor each for: time to station, line name, destination, arrival time… but that seems unwieldy.
It would end up with the following sensors:
1. Arrival, Arrival +1, Arrival +2 (timestamp)
2. Time To Station, Time To Station +1, Time To Station +2 (duration)
3. Line Name, Line Name +1, Line Name +2
4. Destination, Destination +1, Destination +2 

Would it make sense to have this extra information as attributes on an arrival sensor?
So you might have:
1. Arrival, Arrival +1, Arrival +2 - each of these has the following attributes: destination, line name, time to station

Reply:
Adding extra state attributes makes more sense to me so will go that route